glob() is evil in CMake/Py scripts

rolandash opened the pull request:
This PR tries to remove call to glob() from CMake/Py scripts for topic listener.

1. The reason why using glob is evil is already known to popular and can be found at [this post](https://stackoverflow.com/questions/32411963/why-is-cmake-file-glob-evil). In short, when using glob, when list of files changes, CMake can not detect the change of CMakeList file so it will not re-compile. This will cause new added topics not generated automatically and build will fail. 
2. This PR's implementation supports organizing topics/msgs into different folders. This gives more flexibility than just a 2nd external path. As number of topics grows in future extension this can give a more clean view of the structure.

rolandash commented:
thanks @dagar for the hints. 
this commit was made few months ago so i did not notice the print_message change.

I also would like to merge the paths, so the redundancy of set msgfiles can be cut. I am not a CMake expert though more than few times manual reading, not sure how many changes may be required in parent level CMake files. I will ask for your help if i can not get it through.

rolandash commented:
@dagar 
1. seems use CACHE in set variable can extend variable scope to global. The redundancy has been cut now.
2. updated generation py script based on latest commits.

dagar commented on src/modules/micrortps_bridge/CMakeLists.txt:
```diff
@@ -106,7 +106,6 @@ if (GENERATE_RTPS_BRIDGE)
 			-e templates/uorb_microcdr
 			-t ${CMAKE_CURRENT_BINARY_DIR}/tmp/headers_microcdr
 			-q
-		DEPENDS ${msg_files}
```
We need a dependency here so that it's regenerated.

rolandash replied:
Before this commit ${msg_files} is evaluated to empty, as $(msg_files) is a local variable in msg/CMakeList.

dagar replied:
Any ideas how to fix it? Not having dependencies isn't okay because developers will make changes and end up in a state that doesn't make any sense until they do a clean build.

rolandash replied:
I am not very clear about how micrortps is organized, did not have enough time to look into it. 
Looks to me like some bug here, when this command builds OK with an empty dependency, but fail to build with some real value in ${msg_files}.

So what is the original intention to put a dependency here? is it really targeting on make dependency on msg_files, or this variable is just randomly selected as a workaround?

dagar replied:
If you touch a .msg file then these steps need to rerun. Not having a dependency here is even more evil than glob.

dagar commented on src/modules/micrortps_bridge/CMakeLists.txt:
```diff
@@ -123,7 +122,6 @@ if (GENERATE_RTPS_BRIDGE)
 			-e templates/uorb_microcdr
 			-t ${CMAKE_CURRENT_BINARY_DIR}/tmp/sources_microcdr
 			-q
-		DEPENDS ${msg_files}
```
same

dagar commented on src/systemcmds/topic_listener/CMakeLists.txt:
```diff
@@ -32,7 +32,7 @@
 ############################################################################
 
 add_custom_command(OUTPUT topic_listener.cpp
-                COMMAND ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/generate_listener.py ${PX4_SOURCE_DIR} ${EXTERNAL_MODULES_LOCATION} > topic_listener.cpp
```
Why remove EXTERNAL_MODULES_LOCATION?

rolandash replied:
It is not necessary. As now you can set any sub-directory under msg/,   like this:

msg/
       abc.msg
       cde.msg
       external_location/
                        aaa.msg
                        bbb.msg
       other_subdir/
                        ccc.msg

and in msg/CMakeList.txt:

set (msg_files
     ....
      abc.msg
      external_location/aaa.msg
      external_location/bbb.msg
     ....
}

much more flexible for topic expansion. Having an extra EXTERNAL_MODULES_LOCATION param seems redundant.

dagar replied:
External modules is a mechanism for having the entire thing live outside of the tree and get into the build without making changes (environment variable set at configure time).

rolandash replied:
I see.
In that case, will it better to manage to merge the external paths in parent level CMakeFile with $(msg_files) before pass it to py script?

dagar replied:
I believe calling the python script separately would work fine. It's slower to do this for every single .msg file, but doing it in a few groups (msg/, external modules, other_subdir) should be fine.

dagar commented on src/systemcmds/topic_listener/CMakeLists.txt:
```diff
@@ -51,6 +51,7 @@ px4_add_module(
 	SRCS
 		topic_listener.cpp
 	DEPENDS
+		platforms__common
```
Not needed.

rolandash replied:
inherited from previous version. this can be removed.

dagar commented on src/systemcmds/topic_listener/CMakeLists.txt:
```diff
@@ -51,6 +51,7 @@ px4_add_module(
 		generate_topic_listener
 	)
-
+# vim: set noet ft=cmake fenc=utf-8 ff=unix :
```
Not needed. This are used inconsistently, so unless someone wants to fix it entirely we're better off without.

rolandash replied:
this can be removed, inherited from previous version.

dagar commented on msg/CMakeLists.txt:
```diff
@@ -130,6 +130,7 @@ set(msg_files
 	vehicle_status_flags.msg
 	vtol_vehicle_status.msg
 	wind_estimate.msg
+        CACHE STRING "msg files" 
```
indentation

lamping7 commented:
This needs rebasing to overcome past CI issues.

rolandash commented:
Rebased. 
Use set global property instead of CACHE variable, to avoid persisted file list value in cache file across multiple runs of CMAKE, which may cause confusing results.

So a generic pattern this PR implemented for sharing msg file list between different (sub)modules is:
1. in msg/CMakeList a global property is created and maintained to store merged paths for internal msgs and external module msgs(if any).
2. in any other modules who would like to access the list of all msg files, it gets the value from global property using get_property() into a local variable for local usage purpose(like passing to a py script).

I think this pattern can settle all aspects, including build external module orb messages.
@dagar

dagar commented:
Unintentional submodule changes? 
TODO
  - verify touching a .msg is caught by the build (uorb, micrortps, listener)
  - verify external modules are still working

rolandash commented:
Can you be more specific about "Unintentional submodule changes"?

Is external module message build working in old version? I actually doubt that from previous code.

dagar commented:
Take a look at the overall diff.

![image](https://user-images.githubusercontent.com/84712/41547771-1c173096-72ef-11e8-9c8e-67c327e121fc.png)

dagar commented:
> Is external module message build working in old version? I actually doubt that from previous code.

You might be right, what we really need is a simple CI test for it (and everything). Any volunteers?

rolandash commented:
Oh as i rebased the branch in order to pass the CI tests, otherwise always fail during tests. So I updated the origin master from upstream, and then pushed to this commit branch. These submodule updates then shown up here.

rolandash commented:
I can look further into it. That probably requires some changes in serval related python scripts. As that work focus more on external messages, so better in a another separate PR i think.

rolandash commented:
> You might be right, what we really need is a simple CI test for it (and everything). Any volunteers?
I intentionally kept one extra line of code in msg/CMakelist in external message handling part, so the previous behavior when handling external message left unchanged. 
then we can do verify and improvement in another thread.
